IDEA-232541 JUnit5 - Inspect that @Order test method also have correct @TestMethodOrder on class

bedla opened the pull request:
Please check issue https://youtrack.jetbrains.com/issue/IDEA-232541 for detail.

akozlova commented on plugins/junit/src/com/intellij/execution/junit/codeInsight/JUnit5MalformedOrderedTestInspection.kt:
```diff
@@ -0,0 +1,102 @@
+      override fun visitMethod(method: PsiMethod) {
+
+        AnnotationUtil.findAnnotation(method, JUnitCommonClassNames.ORG_JUNIT_JUPITER_API_TEST) ?: return
+        val orderAnnotation = AnnotationUtil.findAnnotation(method, "org.junit.jupiter.api.Order") ?: return
```
please put constants in `JUnitCommonClassNames`

akozlova commented on plugins/junit/src/com/intellij/execution/junit/codeInsight/JUnit5MalformedOrderedTestInspection.kt:
```diff
@@ -0,0 +1,102 @@
+        val psiClass = method.containingClass ?: return
+
+        val testMethodOrderAnnotation = AnnotationUtil.findAnnotation(psiClass, "org.junit.jupiter.api.TestMethodOrder")
+        if (testMethodOrderAnnotation == null) {
```
from the javadoc of `TestMethodOrder`: one can put annotation to the superclass/superinterface, it's not covered inside `AnnotationUtil`

bedla replied:
sorry, I have missed JavaDoc ... it also seems that I have to support `@RepeatedTest`, `@ParameterizedTest`, `@TestFactory`, and `@TestTemplate`

bedla replied:
I also added checks in inheritance hierarchy and more tests (seems that all combinations of inheritance vs annotation and type of testable class is kind of rabbit hole).

akozlova commented on plugins/junit/src/com/intellij/execution/junit/codeInsight/JUnit5MalformedOrderedTestInspection.kt:
```diff
@@ -0,0 +1,102 @@
+  }
+}
+
+private abstract class AnnotationFix : LocalQuickFix {
```
`com.intellij.codeInsight.intention.AddAnnotationFix`?

bedla replied:
i am goning to investigate usage and try to replace it with my `AddAnnotationFix`/`ReplaceAnnotationFix`

bedla replied:
I have replaced my implementations with one you mentioned.

akozlova commented on plugins/junit/test/com/intellij/execution/junit/codeInsight/JUnit5MalformedOrderedTest.kt:
```diff
@@ -0,0 +1,91 @@
+
+  @Throws(Exception::class)
+  override fun setUp() {
+    super.setUp()
```
`com.intellij.execution.junit.codeInsight.JUnit5TestFrameworkSetupUtil`?

bedla commented on plugins/junit/src/com/intellij/execution/junit/codeInsight/JUnit5MalformedOrderedTestInspection.kt:
```diff
@@ -0,0 +1,165 @@
+          holder.registerProblem(
+            problemOnPsiElement,
+            problemDescription,
+            object : RemoveAnnotationQuickFix(testMethodOrderAnnotation, placeOnPsiClass) {
```
Here my tests are green, but when I run fix from UI it only replaces annotation and dont add new annotation - so no replace is occuring from UI, but from tests it does. I am trying to find why but currently without luck (I was looking for some kind of composite quick-fix, but that vararg looks more promising), could you hint me?

bedla commented on plugins/junit/src/com/intellij/execution/junit/codeInsight/JUnit5MalformedOrderedTestInspection.kt:
```diff
@@ -0,0 +1,165 @@
+            problemOnPsiElement,
+            problemDescription,
+            object : RemoveAnnotationQuickFix(testMethodOrderAnnotation, placeOnPsiClass) {
+              // 1st LocalQuickFix's familyName is used to show to the user
```
I am also not sure that when I want to use "composite" quick-fixes with one element in UI that the correct thing is to put familyName on first element. This seems to my that I am doing something wrong.

bedla replied:
i also tried 
```
          holder.registerProblem(
            problemOnPsiElement,
            problemDescription,
            object : LocalQuickFix {
              override fun getFamilyName(): String = "Replace @TestMethodOrder value to support @Order"

              override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
                RemoveAnnotationQuickFix(testMethodOrderAnnotation, placeOnPsiClass).applyFix(project, descriptor)
                AddAnnotationFix(
                  JUnitCommonClassNames.ORG_JUNIT_JUPITER_API_TEST_METHOD_ORDER,
                  placeOnPsiClass,
                  createAnnotationValues(placeOnPsiClass)).applyFix(project, descriptor)
              }
            })
```
but without luck (with `PsiInvalidElementAccessException`)

bedla commented on plugins/junit/src/com/intellij/execution/junit/codeInsight/JUnit5MalformedOrderedTestInspection.kt:
```diff
@@ -0,0 +1,165 @@
+        // but custom implementation of MethodOrderer may take @Order annotation into account (cannot decide)
+      }
+
+      private fun createAnnotationValues(context: PsiClass): Array<out PsiNameValuePair>? {
```
I am also not sure about this creation of `PsiNameValuePair` instances, I mean that creation of artificial annotation from text and then getting list out of it.

bedla commented on plugins/junit/test/com/intellij/execution/junit/codeInsight/JUnit5TestFrameworkSetupUtil.java:
```diff
@@ -69,6 +69,43 @@ public static JavaCodeInsightTestFixture setupJUnit5Library(JavaCodeInsightTestF
                         "public @interface BeforeEach {}");
     fixture.addClass("package org.junit.jupiter.api;\n" +
                         "public interface RepetitionInfo {}");
+    fixture.addClass("package org.junit.jupiter.api;\n" +
```
I can convert this file to kotling for better raw-string readability.

bedla commented:
@akozlova 
I have updated PR to lates master, removed remove-than-add-annotation fix with one add-annotation fix with support of previous annotation removal.
